### PR TITLE
memory: Add 5-level page table tests

### PR DESCRIPTION
**Pull request by kalpanashtty**
* Test will detect CPU feature i.e la57.
      * Check for kernel support
      * Run series of 5-level page tests from "pg-table_tests.git" that covers
            - heap, mmap, shmat tests.

Signed-off-by: Kalpana Shetty <kalpana.shetty@amd.com>

### CODE REVIEW

File: memory/page_table.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        If 5-level page table supported on other platform then\
+        this condition can be removed")
+    def setUp(self):
+        smm = SoftwareManager()
+        for package in ['gcc', 'make']:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        self.url = self.params.get(
+            'url', default="https://github.com/sanskriti-s/pg-table_tests.git")
```
Comment (Naresh-ibm) [inline memory/page_table.py:46]: @kalpanashtty moving this to yaml will help more when we need to change the url.
So suggest to have this in yaml. ok to hard_coded in yaml if needed.

File: memory/page_table.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+           self.cancel("%s is not set in the kernel." % cfg_param)
+        else:
+           self.log.info("Detected 5-Level page table config - CONFIG_X86_5LEVEL set in the kernel")
+
+    def test_pg_table_tests(self):
+        output = process.run('./run-tests')
+        failed_tests = []
+        for line in output.stdout.decode('utf-8').splitlines():
+            if "failed" in line:
+                failed_tests.append(line)
```
Comment (Naresh-ibm) [inline memory/page_table.py:71]: looks like your not adding the failed test name here, so check if we can add the name else having the more meaningful variable name like err = [] or err_msg = [] etc.

File: memory/page_table.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    def test_kernel(self):
+        cfg_param = "CONFIG_X86_5LEVEL"
+        result = linux_modules.check_kernel_config(cfg_param)
+        if result == linux_modules.ModuleConfig.NOT_SET:
+           self.cancel("%s is not set in the kernel." % cfg_param)
+        else:
+           self.log.info("Detected 5-Level page table config - CONFIG_X86_5LEVEL set in the kernel")
+
+    def test_pg_table_tests(self):
```
Comment (Naresh-ibm) [inline memory/page_table.py:66]: @kalpanashtty add doc_strings for all the methods.

**Review by Naresh-ibm — CHANGES_REQUESTED** [review]

File: memory/page_table.py (hunk 1 above)
Comment (kalpanashtty) [reply memory/page_table.py:46]: Actually, I forgot to create yaml folder... will do.

File: memory/page_table.py (hunk 2 above)
Comment (kalpanashtty) [reply memory/page_table.py]: Done

File: memory/page_table.py (hunk 3 above)
Comment (kalpanashtty) [reply memory/page_table.py:76]: Added comments.

File: memory/page_table.py (hunk 1 above)
Comment (kalpanashtty) [reply memory/page_table.py:50]: Done

File: memory/page_table.py (hunk 3 above)
Comment (kalpanashtty) [reply memory/page_table.py:76]: Done

File: memory/page_table.py (hunk 1 above)
Comment (Naresh-ibm) [reply memory/page_table.py:50]: adding the hard_coded value for default url is not required here as we are already doing it in yaml. 
so please remove this and add `default=None`

File: memory/page_table.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        git.get_repo(self.url, destination_dir=self.teststmpdir)
+        os.chdir(self.teststmpdir)
+        build.make(self.teststmpdir)
+
+    def test_detect_5lvl(self):
+        '''
+        Detect 5-Level page table CPU feature
+        '''
+        cpu_info = genio.read_file("/proc/cpuinfo")
+        if 'la57' in cpu_info:
```
Comment (Naresh-ibm) [inline memory/page_table.py:60]: I feel this should be exact match and may give the false True here if this is the part of some other string.
so please re-check. I am ok if you think this is a unique string.

**Review by Naresh-ibm — CHANGES_REQUESTED** [review]

File: memory/page_table.py (hunk 4 above)
Comment (kalpanashtty) [reply memory/page_table.py:60]: Yes, it's a unique string in "/proc/cpuinfo" for 5-Level page table support.
Ref: https://lwn.net/Articles/716324/

**Comment by Naresh-ibm** [comment]
@kalpanashtty please add the job/debug.log for reference along with run logs.

File: memory/page_table.py (hunk 1 above)
Comment (kalpanashtty) [reply memory/page_table.py]: > adding the hard_coded value for default url is not required here as we are already doing it in yaml. so please remove this and add `default=None`

Done.

**Comment by kalpanashtty** [comment]
> @kalpanashtty please add the job/debug.log for reference along with run logs.

Attached...
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9751763/job.log)

[1-page_table.py_PageTable.test_detect_5lvl_run-6e47_debug.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9751765/1-page_table.py_PageTable.test_detect_5lvl_run-6e47_debug.log)

[2-page_table.py_PageTable.test_kernel_run-6e47_debug.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9751768/2-page_table.py_PageTable.test_kernel_run-6e47_debug.log)

[3-page_table.py_PageTable.test_pg_table_tests_run-6e47_debug.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/9751769/3-page_table.py_PageTable.test_pg_table_tests_run-6e47_debug.log)

**Comment by kalpanashtty** [comment]
root@xxxxx:/home/xxxxx/avocado-misc-tests/memory# avocado run --test-runner runner page_table.py -m /home/xxxxx/avocado-misc-tests/memory/page_table.py.data/page_table.yaml
JOB ID     : 5ed873c544465f89a518bc602b67d7da02a2c5f9
JOB LOG    : /home/xxxx/tests/results/job-2022-10-11T04.53-5ed873c/job.log
 (1/3) page_table.py:PageTable.test_detect_5lvl;run-6e47: PASS (2.19 s)
 (2/3) page_table.py:PageTable.test_kernel;run-6e47: PASS (1.72 s)
 (3/3) page_table.py:PageTable.test_pg_table_tests;run-6e47: PASS (2.77 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/xxxxx/tests/results/job-2022-10-11T04.53-5ed873c/results.html
JOB TIME   : 7.68 s

**Comment by Naresh-ibm** [comment]
Thanks @kalpanashtty 
LGTM

**Comment by kalpanashtty** [comment]
@narasimhan-v  @PraveenPenguin:  @Naresh-ibm  has done the review, so, whenever you have time please review it.

**Review by narasimhan-v — COMMENTED** [review]
@kalpanashtty Just a minor question / suggestion. Will `test_detect_5lvl()` run if `test_detect_5lvl()` and `test_detect_5lvl()` cancels ?
If they don't run in that case, we need to move the 2 functions into `setUp()`

**Comment by kalpanashtty** [comment]
> @kalpanashtty Just a minor question / suggestion. Will `test_detect_5lvl()` run if `test_detect_5lvl()` and `test_detect_5lvl()` cancels ? If they don't run in that case, we need to move the 2 functions into `setUp()`

Yes; it would run other tests... primarily, adding manual tests as different test cases so that if the cpu feature "la57" is present it would pass. So, thinking we can make the test fail instead of cancel might give more meaning.

But somehow adding it as part of "setUp()" will not differentiate different test cases. Also, in the future, we will add additional test cases and that would repeat 'setUp' for each of those tests.
So, in my opinion, if we can make the test fail would make more sense instead of adding 1st 2 cases as part of 'setUp'.

**Comment by kalpanashtty** [comment]
> > @kalpanashtty Just a minor question / suggestion. Will `test_detect_5lvl()` run if `test_detect_5lvl()` and `test_detect_5lvl()` cancels ? If they don't run in that case, we need to move the 2 functions into `setUp()`
> 
> Yes; it would run other tests... primarily, adding manual tests as different test cases so that if the cpu feature "la57" is present it would pass. So, thinking we can make the test fail instead of cancel might give more meaning.
> 
> But somehow adding it as part of "setUp()" will not differentiate different test cases. Also, in the future, we will add additional test cases and that would repeat 'setUp' for each of those tests. So, in my opinion, if we can make the test fail would make more sense instead of adding 1st 2 cases as part of 'setUp'.

Discussed with Narasimhan and we will make test fail if we don't see system is capable of 5-level feature instead cancel the tests. So pushed the changes for review.

**Review by PraveenPenguin — APPROVED** [review]

**Review by narasimhan-v — APPROVED** [review]

**Review by Naresh-ibm — APPROVED** [review]
@PraveenPenguin @narasimhan-v Thanks for reviews
@kalpanashtty Thanks for accommodating all the changes suggested.

**Pull request merged**